waiter/src/waiter/interstitial.clj

```diff
@@ -0,0 +1,237 @@
+                                                    (fn [service-id->interstitial-promise]
+                                                      (pc/map-vals #(deref % 0 :not-realized) service-id->interstitial-promise)))
+                              :maintainer current-state})
+                           (async/>! response-chan))
+                      current-state)
+
+                    scheduler-state-chan
+                    (let [scheduler-messages message
+                          available-service-ids' (process-scheduler-messages
+                                                   interstitial-state-atom service-id->service-description
+                                                   available-service-ids scheduler-messages)]
+                      (when-not (:initialized? @interstitial-state-atom)
+                        (log/info "interstitial state has been initialized with services from the scheduler")
+                        (swap! interstitial-state-atom assoc :initialized? true))
+                      (assoc current-state :available-service-ids available-service-ids'))))
+                (catch Exception e
+                  (log/error e "error in interstitial-maintainer")
+                  current-state))]
+          (when next-state
+            (recur next-state)))))
+    {:exit-chan exit-chan
+     :query-chan query-chan}))
+
+(let [interstitial-template-fn (template/fn
+                                 [{:keys [service-description service-id target-url]}]
+                                 (slurp (io/resource "web/interstitial.html")))]
+  (defn render-interstitial-template
+    "Renders the interstitial html page."
+    [context]
+    (interstitial-template-fn context)))
+
+(def ^:const bypass-interstitial-param-name-value "x-waiter-bypass-interstitial=1")
+
+(defn wrap-interstitial
+  "Redirects users to the interstitial page when interstitial has been enabled and the service has not been started.
+   The interstitial page will retry the request as a GET with the same query parameters but bypass the interstitial page."
+  [handler interstitial-state-atom store-service-description]
+  (fn wrap-interstitial-handler [{:keys [descriptor headers query-string uri] :as request}]
+    (let [{:keys [service-description service-id]} descriptor
+          {:strs [interstitial-secs]} service-description
+          ;; the bypass interstitial should be the last query parameter
+          bypass-interstitial? (str/ends-with? (str query-string) bypass-interstitial-param-name-value)]
+      (if (or bypass-interstitial? ;; bypass query parameter provided
+              (zero? interstitial-secs) ;; interstitial support has been disabled
+              (not (str/includes? (str (get headers "accept")) "text/html")) ;; expecting html response
+              (headers/contains-waiter-header headers sd/service-description-keys) ;; on-the-fly request
```

Comment:
Should we add `:on-the-fly?` to the descriptor?  We're already performing this check in `compute-service-description`.  Also we likely want to use `sd/on-the-fly-service-description-keys`.  

This will be called on every request for interstitial-enabled services. It might be preferable to do it only once inside `compute-service-description`.

Reply:
Added `:on-the-fly?`